File: w10l4.cpp
//OBSERVER PATTERN-establishes a one to many relationship between objects
// #include<iostream>
// #include<string>
// #include<vector>
// class Observer{
//     public:
//     virtual void update(const std::string& message)=0;

// };
// class ConcreteObserver:public Observer{
//     public:
//     void update(const std::string& message)override{
//         std::cout<<"Observer received: "<<message<<std::endl;
//     }

// };
// class Subject{
//     std::vector<Observer*> observers;
//     public:
//     void attach(Observer* observer){
//         observers.push_back(observer);
//     }
//     void notify(const std::string& message ){
//         for(auto observer:observers){
//             observer->update(message);
//         }
//     }
// };
// int main(){
//     Subject subject;
//     ConcreteObserver obs1 ,obs2;
//     subject.attach(&obs1);
//     subject.attach(&obs2);
//     subject.notify("Hello!!");
//     return 0;
// }

// #include<iostream>
// #include<vector>
// #include<string>
// //Abstract class
// class Observer{
//     public:
//     virtual void update(float temperature,float humidity)=0;
// };
// class Phonedisplay:public Observer{
//     public:
//     void update(float temperature,float humidity)override{
//         std::cout<<"TEMP- "<<temperature<<" Degree- "<<humidity<<std::endl;
//     }
// };
// class Weatherstation{
//     std::vector<Observer* >observers;
//     float temperature,humidity;
//     public:
//     void attach(Observer* observer){
//         observers.push_back(observer);
//     }
//     void setm(float temp,float hum){
//         temperature=temp;
//         humidity=hum;
//         notify();
//     }
//     void notify(){
//         for (auto observer:observers){
//             observer->update(temperature,humidity);
//         }
//     }
// };
// int main(){
//     Weatherstation station;
//     Phonedisplay phone;
//     station.attach(&phone);
//     station.setm(25.0,60.0);
//     station.setm(30.0,55.0);
//     return 0;
// }

File: w10l4.java
import java.util.ArrayList;
import java.util.List;
class Subject{
    private List<Observer> observers=new ArrayList<>();
    public void attach(Observer observer){
        observers.add(observer);
    }
    public void notify(String message){
        for(Observer observer:observers){
            observer.update(message);
        }
    }
}
interface Observer{
    void update(String message);
}
class ConcreteObserver implements Observer{
    public void update(String message){
        System.out.println("Observer received: "+message);
    }
}
public class w10l4 {
    public static void main(String[] args) {
        Subject subject=new Subject();
        Observer ob1=new ConcreteObserver();
        Observer ob2=new ConcreteObserver();
        subject.attach(ob1);
        subject.attach(ob2);
        subject.notify("Hello Observers!!");
    }
    
}
